File: Data Structures & Algorithms/search-2d-matrix/submission-2.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {

        int row;
        for (row = 0; row<matrix.size()-1; row++) {
            if (matrix[row][0] <= target && target < matrix[row+1][0]) {
                break;
            }
        }

        for (int num:matrix[row]) {
            if (num == target) return true;
        }

        return false;
    }
};

File: Data Structures & Algorithms/perfect-squares/submission-2.py
class Solution:
    def numSquares(self, n: int) -> int:
        dp = [n]*(n+1)
        dp[0] = 0
        for num in range(1, n+1):
            s = 1
            while s ** 2 <= num:
                dp[num] = min(dp[num], 1 + dp[num-s**2])
                s += 1

        return dp[n]